internal/controllers/addon/monitoring_stack_reconciler.go

```diff
@@ -0,0 +1,225 @@
+package addon
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	monv1 "github.com/rhobs/obo-prometheus-operator/pkg/apis/monitoring/v1"
+	obov1alpha1 "github.com/rhobs/observability-operator/pkg/apis/monitoring/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8sApiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"github.com/openshift/addon-operator/internal/controllers"
+)
+
+const MONITORING_STACK_RECONCILER_NAME = "monitoringStackReconciler"
+
+var errMonitoringStackSpecNotFound = fmt.Errorf("monitoring stack spec not found")
+
+type monitoringStackReconciler struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func (r monitoringStackReconciler) Name() string {
+	return MONITORING_STACK_RECONCILER_NAME
+}
+
+func (r *monitoringStackReconciler) Reconcile(ctx context.Context,
+	addon *addonsv1alpha1.Addon) (ctrl.Result, error) {
+
+	// ensure creation of MonitoringStack object
+	latestMonitoringStack, err := r.ensureMonitoringStack(ctx, addon)
+	if err != nil {
+		if errors.Is(err, errMonitoringStackSpecNotFound) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// propagate the recently reconciled (created/updated) monitoring stack's status to the owner Addon
+	if monitoringStackAvailable := r.propagateMonitoringStackStatusToAddon(latestMonitoringStack, addon); !monitoringStackAvailable {
+		return handleExit(resultRetry), nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *monitoringStackReconciler) ensureMonitoringStack(ctx context.Context,
+	addon *addonsv1alpha1.Addon) (*obov1alpha1.MonitoringStack, error) {
+	if !HasMonitoringStack(addon) {
+		return nil, errMonitoringStackSpecNotFound
```

Comment:
ahh good catch, yeah that shouldn’t happen.

Reply:
hmm but we just return `ctrl.Result{}, nil` in other parts of ADO when the object the sub-reconciler works on doesn't exist. 
Also gorm and k8s client libs returning an error when getting a resource is different from this i feel. In those cases you asked for a resource with a specific ID/Name/X so if it doesn't exist that counts as an error.
wdyt?

Reply:
> hmm but we just return ctrl.Result{}, nil in other parts of ADO when the object the sub-reconciler works on doesn't exist.

Yeah I wanted to personally steer away from this as I was not very fond of the existing implementation. Not that the existing implementation is wrong or inoperable, just semantically speaking, returning an ErrNotFound made more idiomatic sense considering it's only providing more information to the caller and giving it the opportunity to take a more informed decision of why to return a `ctrl.Result{}, nil`.

The return values of `ensureMonitoringStack` shouldn't be determined on the potential intent or use-case of the caller but it should be based on the intent of what `ensureMonitoringStack` wants to do which is to "ensure the existence of MonitoringStack". And from that perspective, it clearly makes sense that if it tries to ensure the existence of a "non-existent object" (`if !HasMonitoringStack(addon)`), it should throw an error. 

It is analogous to the controller-runtime throwing an error when we try to build the scheme with a non-existent GVK or k8s/gorm returning a NotFound error, not just on a GET, but even on UPDATES, when acted upon a non-existent object?

But even if we ignore the examples of controller-runtime, gorm, etc. even idiomatically, as I justified in the first section, from just the POV of the `ensureMonitoringStack` and not the caller, it makes more sense to throw a `ErrNotFound` error instead of returning success.